File: java/EncodeandDecodeStrings.java
/*
 Design an algorithm to encode a list of strings to a single string.
 The encoded string is then decoded back to the original list of strings.
 
 Example 1:
 
Input: ["neet","code","love","you"]
Output:["neet","code","love","you"]
 
 Example 2:
 Input: ["we","say",":","yes"]
 Output: ["we","say",":","yes"]
 
 */

import java.util.ArrayList;
import java.util.List;

 class Solution {

    // Encodes a list of strings to a single string.
    public String encode(List<String> strs) {
        StringBuilder encodedString = new StringBuilder();
        for (String str : strs) {
            // use the length of the string followed by a '#' to indicate the length of the string
            encodedString.append(str.length()).append('#').append(str);
        }
        return encodedString.toString();
    }

    // Decodes a single string to a list of strings.
    public List<String> decode(String s) {
        List<String> decodedList = new ArrayList<>();
        int i = 0;
        while (i < s.length()) {

            // Find the position of the delimiter '#'
            int j = s.indexOf('#', i);

            // Get the length of the next string
            int length = Integer.parseInt(s.substring(i, j));

            // Move the index to the start of the string
            i = j + 1;

            // Extract the string based on the length
            String str = s.substring(i, i + length);
            decodedList.add(str);
            
            // Move the index forward by the length of the string
            i += length;
        }
        return decodedList;
    }
}

File: python/Encode and Decode Strings.py
# Encode and Decode Strings - 271

class Solution:

    # Encodes a list of strings into a single string
    def encode(self, strs: List[str]) -> str:
        # Use a generator expression to create a string for each input string
        # The string is in the format "{length}#{string}", where length is the length of the string
        return ''.join(f"{len(s)}#{s}" for s in strs)

    # Decodes a single string back into a list of strings
    def decode(self, s: str) -> List[str]:
        # Initialize an empty list to store the decoded strings
        strs = []
        # Initialize a pointer to the current position in the string
        i = 0
        # Loop until we've processed the entire string
        while i < len(s):
            # Find the position of the '#' character, which separates the length from the string
            j = s.find('#', i)
            # Extract the length of the string from the substring before the '#'
            length = int(s[i:j])
            # Move the pointer past the '#' character
            i = j + 1
            # Extract the string itself from the substring of the specified length
            strs.append(s[i:i+length])
            # Move the pointer past the extracted string
            i += length
        # Return the list of decoded strings
        return strs